cabal-install/Distribution/Client/FileStatusCache.hs

```diff
@@ -0,0 +1,819 @@
+  deriving (Show, Generic)
+
+instance Binary FileGlobMonitorState
+
+-- We can build a 'MonitorStateFileSet' from a set of 'MonitorFilePath' by
+-- inspecting the state of the file system, and we can go in the reverse
+-- direction by just forgetting the extra info.
+--
+reconstructMonitorFilePaths :: MonitorStateFileSet -> [MonitorFilePath]
+reconstructMonitorFilePaths (MonitorStateFileSet singlePaths globPaths) =
+    Map.foldrWithKey (\k x r -> getSinglePath k x : r)
+                     (map getGlobPath globPaths)
+                     singlePaths
+  where
+    getSinglePath filepath monitorState =
+      case monitorState of
+        MonitorStateFile{}          -> MonitorFile            filepath
+        MonitorStateFileHashed{}    -> MonitorFileHashed      filepath
+        MonitorStateFileNonExistent -> MonitorNonExistentFile filepath
+        MonitorStateFileChanged     -> MonitorFile            filepath
+        MonitorStateFileHashChanged -> MonitorFileHashed      filepath
+
+    getGlobPath (MonitorStateGlobDirs  glob globs _ _) =
+      MonitorFileGlob (GlobDir  glob globs)
+    getGlobPath (MonitorStateGlobFiles glob       _ _) =
+      MonitorFileGlob (GlobFile glob)
+
+------------------------------------------------------------------------------
+-- Checking the status of monitored files
+--
+
+-- | A monitor for detecting changes to a set of files. It can be used to
+-- efficiently test if any of a set of files (specified individually or by
+-- glob patterns) has changed since some snapshot. In addition, it also checks
+-- for changes in a value, and when there are no changes in either it returns
+-- a saved value.
+--
+-- The main use case looks like this: suppose we have some expensive action
+-- that depends on certain pure inputs and reads some set of files, and
+-- produces some pure result. We want to avoid re-running this action when it
+-- would produce the same result. So we need to monitor the files the action
+-- looked at, the other pure input values, and we need to cache the result.
+-- Then at some later point, if the input value didn't change, and none of the
+-- files changed, then we can re-use the cached result rather than re-running
+-- the action.
+--
+-- This can be achieved using a 'FileMonitor'. Each 'FileMonitor' instance
+-- saves state in a disk file, so the file for that has to be specified,
+-- making sure it is unique. The pattern is to use 'checkFileMonitorChanged'
+-- to see if there's been any change. If there is, re-run the action, keeping
+-- track of the files, then use 'updateFileMonitor' to record the current
+-- set of files to monitor, the current input value for the action, and the
+-- result of the action.
+--
+-- The typical occurrence of this pattern is captured by 'rerunIfChanged'
+-- and the 'Rebuild' monad. More complicated cases may need to use
+-- 'checkFileMonitorChanged' and 'updateFileMonitor' directly.
+--
+data FileMonitor a b
```

Comment:
The comment should probably mention what the `b` parameter is for. A phantom type I presume?